Container documentation

CemTunaboylu opened the pull request:
Linux filesystem coverage is finished. There are a lot of details, but this was barely scratching the surface. I think I delivered enough, maybe even more. This may need more formatting, definitely needs feedback.

CemTunaboylu committed:
The bare back bone of the container

go run main.go run <cmd> [args]

Now, if you are a root user you can directly execute any command. If not
you should sudo it first. There is nothing here now that isolates the
process from the host except the Unix Timesharing System which creates
an isolation around the hostname only.

CemTunaboylu committed:
Cgroups addition for process max

Mounting is stil on hold. We need Ubuntu filesystem for that.

CGroups do not work - as far as I am aware - with rootless containers.
For now, running as a root user 'disarms' the fork bomb.

`main.go` in this setting is the backbone of the container. There is
nothing that I have added now.

CemTunaboylu committed:
Cgroup SS file name correction

CemTunaboylu committed:
Somehow initialized

Here, we built on Liz Rice's sequel. I broadly described what I am going
to do. From here on, I 'take on', will refactor the code, and add some
tests only so that we see this backbone works. After test inclusions,
containerization will not be worked on until the image part is done.

The project architecture is NOT established yet. After the design
processes finishes, and the project frame is visible, then it will be
decided.

CemTunaboylu committed:
Segregating the documentation from code

One of the most important parts is unraveling the mystery of containers
and images. Now, the code does not contribute to anything for us to understand. I segregated this part for this purpose. Without getting to a certain point in this branch, other branches will wait. This branch not only will include the README and TODOs but also the deep diving all the processes that we will go through. Like a surgeon, I plan to operate on todays technologies and how/what they are doing with the best resolution I can get into. After we have a deep understanding of the underlying the technologies, every code will make sense. So I think this branch is to be the starting point.

CemTunaboylu committed:
Started in Linux Filesystem Deep Dive

I finished the necessary filesystem coverage for Linux.

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,352 @@
+# Deep Diving 
+
+Before starting, I want to clear some of my notations. I will abide to the conventions of Linux man pages. In those:
```
> I want to clear

I want to *clarify*

ozgunozerk replied:
> I will abide to

I will abide *by*

CemTunaboylu replied:
Hahahahah incredibly appreciated 👼🏻

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,352 @@
+
+Before starting, I want to clear some of my notations. I will abide to the conventions of Linux man pages. In those:
+
+`[ ]` means that anything within (any values, information or parameters) are optional. You can chose one, more than one or none at all.
```
> You can chose

You can *choose*

ozgunozerk replied:
> (any values, information or parameters)

(any values, information`,` or parameters

> anything within (any values, information or parameters) are optional

anything within (any value, information, or parameter) *is* optional (anything is singular, so the parenthesis should be singular as well)

CemTunaboylu replied:
Yes!! I should have take a step back and read the sentences again ...

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,352 @@
+`< >` means that the element within (again any parameter, information or a value) is mandatory. We require to replace the text in between angle brackets with actual appropriate information.
+`-`means the options. It does not stand by itself, it will be followed by some form of a value ( or information or parameter).
+
+`-` We will not use that one I guess but while on it a little extra does not kill anyone ... It means long options like a more descriptive name of the option name above. It is pretty common and you may stumble upon them anywhere. For example an option `-l` indicating an option as a 'list', will be given `--list` as a long option as well. Both accomplish the same thing. When you use heavily on opt ...
```
> For example 

For example, *(comma is a must), also see line 11 for the same issue. Probably there are more, you can ctrl+f and search the whole document 🗡️*

ozgunozerk replied:
> We will not use that one I guess but while on it a little extra does not kill anyone

I suggest re-writing this sentence. Too much conjunctions:
- I guess
- but
- while

also, *while on it* is a bit unclear. Maybe *while we're on it*, but I still suggest rewriting this from scratch.

ozgunozerk replied:
> It means long options

*It is the shorthand version of the long options* 
Is it what you mean? If so, `it means long options` actually confused me :D

CemTunaboylu replied:
> > We will not use that one I guess but while on it a little extra does not kill anyone
> 
> I suggest re-writing this sentence. Too much conjunctions:
> 
> * I guess
> * but
> * while
> 
> also, _while on it_ is a bit unclear. Maybe _while we're on it_, but I still suggest rewriting this from scratch.

Yes you are absolutely right. I am rewriting and dividing it to parts.

CemTunaboylu replied:
> > It means long options
> 
> _It is the shorthand version of the long options_ Is it what you mean? If so, `it means long options` actually confused me :D

I mean normal options with 1 `–` are shorthand versions. If you use `--` you explicitly write the whole options name. Example: `-l` and `--list`. What I meant was pretty obscure, I am changing that too.

ozgunozerk replied:
Yes I understood it correctly then. You confused me on something I already know haha

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,352 @@
+
+`-` We will not use that one I guess but while on it a little extra does not kill anyone ... It means long options like a more descriptive name of the option name above. It is pretty common and you may stumble upon them anywhere. For example an option `-l` indicating an option as a 'list', will be given `--list` as a long option as well. Both accomplish the same thing. When you use heavily on opt ...
+
+`{ }` I am not sure if this is common. This means within are required options that can be in any relation such as mutual exclusion and so on. One of the specified options or some in conjunction to others *must* be present. For example filtring some results with `{ --time $TODAY | --priority 9 }` means "I want the ones which are from today( assume that entries have times associated with them)" **O ...
```
> conjunction to others

conjunction *with* others

ozgunozerk replied:
> If none of them are satisfied

If none of them *is* satisfied

ozgunozerk replied:
> returned result is empty

*the* returned result is empty

CemTunaboylu replied:
Thank you 🦅 you have eyes of an eagle ...

ozgunozerk replied:
It is grammarly, not me :)

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,352 @@
+Before starting, I want to clear some of my notations. I will abide to the conventions of Linux man pages. In those:
+
+`[ ]` means that anything within (any values, information or parameters) are optional. You can chose one, more than one or none at all.
+`< >` means that the element within (again any parameter, information or a value) is mandatory. We require to replace the text in between angle brackets with actual appropriate information.
```
> We require to replace

We require *replacing*

CemTunaboylu replied:
Fixed it

ozgunozerk reviewed (commented):
First iteration on being a grammar nazi 🙊

ozgunozerk commented:
Absolutely loved the memes! Wonderful addition 🎉

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,352 @@
+
+Metadata of `images/`             |  metadata of `README.md`
+:-------------------------:|:-------------------------:
+![metadata of folder](/images/metadata_of_images.png)  | ![metadata of README.md file](/images/metadata_of_md.png)
```
It is actually amazing to see the differences. However, it is a bit hard for a human eye to compare them. It took me 20 seconds to realize their first line was the same entry... 
I thought, the left one was missing some fields as the first entries, compared to the right one. 

But for the later entries, they are aligned, and it is very easy to compare. So if you align top entries, you will have to make a compromise from the alignment of the later entries. What I suggest may seem like a lot of work, but I think it will make it way easier for the reader to understand what's going on.

The suggestion:
Divide these images into 2 images (in total, you will have 4 images). The first 2 will be for aligning the top entries, the last 2 will be aligning the rest of the entries.

CemTunaboylu replied:
This makes sense. On top of that, I should color same ones and different ones for the sake of the reader I guess. Thank you for this input.

CemTunaboylu replied:
I changed it a bit. How is it now?

ozgunozerk replied:
wonderful, simpler yet effective :)

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,352 @@
+`/` *refers to the root directory. Root directory is - as its name suggests - the directory from which all the other directories branches off.* 
+), my life will flash before my eyes. 
+
+![Without limited tree cmd](/images/tree_without_limits.mov)
```
seems like movies cannot be rendered in md files :(

CemTunaboylu replied:
I will try to drag and drop which seems to be valid in new git. If I am able to make this work, I will explain it so that we can use it, I think it may be useful. 🤓

CemTunaboylu replied:
I fixed it. Have a separate commit 7900b14  for this one, you may check it out.

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,352 @@
+![metadata of folder](/images/metadata_of_images.png)  | ![metadata of README.md file](/images/metadata_of_md.png)
+
+
+You can see that they are not awfully different. It is trivial to see know that a directory is a file itself that can comtain multiple other files.
```
> It is trivial to see know that

It is trivial to see that

ozgunozerk replied:
> comtain

contain

CemTunaboylu replied:
> > comtain
> 
> contain

I see that my fingers let me down on this one. 

> > It is trivial to see know that
> 
> It is trivial to see that

But WTF is this one? Where does the know come from?

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,352 @@
+
+You might be wondering why there is an arrow `->` after **`/bin`**. `tree` shows symbolic links in that format : `name -> real-path`. A symbolic link - you may have encountered it referred as a symlink, shell link or a soft link - points to another file or directory on the system or a connected file system. A fancy way of saying a `shortcut`. So my **`/bin`** directory actually points to **`usr/b ...
+
+You see yourself where that programs or commands reside with `which`. For example, the command to move stuff around `mv`:
```
> You see yourself

did you mean -> *you see for yourself* ?
if so, you can simply get rid of `yourself` as well. Does not add much value to the content imo

CemTunaboylu replied:
I don't know even remember what I meant ... Not this one for sure.

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,352 @@
+
+![lsa_boot](/images/lsa_boot.png)
+
+**`/boot`** includes what your system needs when starting a.k.a. booting. There are information for GRUB (GRand Unified Bootloader) - the boot loader package - and all the kernels - we will dive deeper into kernels, but for now, it is basically the core program of OS that has complete control over everything in the system - you have installed, their configuration files and so on. 
```
> when starting a.k.a. booting.

when starting (a.k.a. booting).

ozgunozerk replied:
> over everything in the system - you have installed, their configuration files and so on.

*over everything in the system (you have installed, their configuration files and so on).*
The reason is, you used the `-` separator as giving a sneak peak on a divergent topic. But here, it is actually the explanation of the previous statement, hence, I suggest a different syntax. I think parenthesis will do the job!

CemTunaboylu replied:
> > over everything in the system - you have installed, their configuration files and so on.
> 
> _over everything in the system (you have installed, their configuration files and so on)._ The reason is, you used the `-` separator as giving a sneak peak on a divergent topic. But here, it is actually the explanation of the previous statement, hence, I suggest a different syntax. I think parenthesis will do the job!

Thank you for this one especially. Usage of `-` and `(`seems inconsistent throughout the document. Do you have any suggestions for separating their use as separate duties they will have?

ozgunozerk replied:
following standard language rules always worked out for me. 
`-` is for elaborating on a divergent topic
`(` is for giving some extra definition/explanation on the topic, that you think may not be understood well by the audience.

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,352 @@
+
+**`/boot`** includes what your system needs when starting a.k.a. booting. There are information for GRUB (GRand Unified Bootloader) - the boot loader package - and all the kernels - we will dive deeper into kernels, but for now, it is basically the core program of OS that has complete control over everything in the system - you have installed, their configuration files and so on. 
+
+It is a very important directory, it is not a good idea to touch any of these...
```
two times `it is` in a sequence is not a feast on the eyes.

Maybe paraphrase as: *It is a very important directory, and wouldn't be wise to touch any of these...*

CemTunaboylu replied:
Thank you, I changed it.

CemTunaboylu committed:
Changes/Fixes according to the PRs.

Now, only the changes suggested are performed. After this push, whole
document will be introduced to a text/grammar checker. It will be a
separate commit for clarity's sake.

The video does not work. It will be fixed on github web.

CemTunaboylu committed:
Grammarly Check + Corrections

It is more useful than I thought.

CemTunaboylu committed:
Metadata comparison for README.md and /images directory changed

I took an SS. Than colored where they start is convergent and how they are not aligned
to indicate otherwise.  I mentioned this explicitly. Further
improvements can be made.

CemTunaboylu committed:
Movie embed

At line 43, there was an attempt to embed a video -> it has failed. GitHub seems to have an improvement on this. If you drag and drop the `.mov` file, it uploads it to a cloud location and references it there.

CemTunaboylu committed:
Delete `tree_withoout_limits.mov`

It does not serve any purpose.

CemTunaboylu committed:
Subspace addition

I could not help my self.

CemTunaboylu committed:
minor fixes

typo fixes

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,350 @@
+
+---
+
+**`/lost+found`** is as its name suggests for data fragments there are not referenced anywhere anymore. Let us step back for a second and see it clearly. When your system is not properly shut down, due to a crash or a failure, the next boot a diligent file system check using `fsck` will take place. Wait, what the hell is `fsck`? `fsck` is a system utility tool that helps detect inconsistencies of ...
```
> When your system is not properly shut down, due to a crash or a failure, the next boot a diligent file system check using `fsck` will take place.

due to a crash or a failure, **at** the next boot

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,350 @@
+
+---
+
+**`/mnt`** is a geezer one too. It was used to manually mount storage, and temporarily mount a filesystem as needed. This directory does not affect any execution or run of any program. If I am to farm in Supspace network for example, to increase my memory and to gain more tickets, I add an SSD to my system, I can set up a permanent mount point for it here.
```
:)

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,350 @@
+
+![ls root](/images/lsa_root.png)
+
+**`/root`** is equivalent to `root` - a.k.a. administrator or super - users **`/home`**. 
```
it seemed to me, these are the keywords that can help you to remember the topic if you are giving a presentation. Maybe you can elaborate a bit more :)
An outsider to the Linux filesystem probably won't understand what you mean here

CemTunaboylu replied:
You mean `root` and `admin`? These are so engrained in life, I did not notice that I did not touch on them! Thank you, this is VERY precious.

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,350 @@
+
+![lsa snap ](/images/lsa_snap.png)
+
+**`/snap`** is for storing your favorite Snapchat stories. I am kidding. Let us read what the `README` says within the **`/snap`** directory. 
```
actually laughed out loud! great refresher 🤣

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,350 @@
+
+![snap readme](/images/snap_readme.png)
+
+Let me refresh your memory: `apt` was a cli handling installation or removing the packages. It is a packaging tool. 'snap' - not the directory, the program - is similar, it is a packaging and deployment system. Snaps are a secure and scalable way to embed applications on Linux devices. A snap is an application **containerized** with all its dependencies. A snap can be installed using a single com ...
```
Thought, there was an incident, where in a snap, the developer put cryptocurrency mining software and all hell broke loose...

did not understand the meaning of `Thought` in the beginning. Was it `think`, or was it `though`?

CemTunaboylu replied:
though 🥲 Apparently, I don't know what I am writing while writing.

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,350 @@
+
+Let me refresh your memory: `apt` was a cli handling installation or removing the packages. It is a packaging tool. 'snap' - not the directory, the program - is similar, it is a packaging and deployment system. Snaps are a secure and scalable way to embed applications on Linux devices. A snap is an application **containerized** with all its dependencies. A snap can be installed using a single com ...
+
+Those packages are also called snaps. Snaps are self-contained applications running in a sandbox with mediated access to the host system. Self-contained means they consist of a single, installable bundle containing the application and a copy of the run-time environment needed to run the application. When the application is installed, it behaves in the same way as any native application. The file  ...
```
> Those packages are also called snaps

did not understand what does this `those packages` actually refer to :(

ozgunozerk replied:
How come snaps use less disk space? Less disk space compared to normal applications (non-sandboxed)? If so, I don't get it. Snaps include: `the original application + the runtime`, doesn't it? If so, they cannot be smaller than a normal application?

CemTunaboylu replied:
This part was a bummer, I am glad that it didn't slip. When I was writing I dumped everything, and pasted it together. I was not sure what to do here. Snaps are zip files. Their data is zipped. On the fly, snapd unzips them. They also share common dependencies. I wrote they use less space, but I am not sure if this holds because what do we compare them to? Common dependencies should be shared among anything practically in the OS as long as I am aware. SO I thought I deleted this part but it turns out it slipped in. What do you think?

ozgunozerk replied:
Yes, I think deleting that part and not confusing the reader is a better choice. This is not about snaps in the end. It was a detail 👍

ozgunozerk commented on README.md:
```diff
@@ -1,2 +1,350 @@
+
+---
+
+**`/srv`** contains data for the systems provided services. This includes your servers. Let us say we are running an FTP server
```
> contains data for the systems provided services

contains `the` data for the system`'`s provided services

ozgunozerk commented:
What a wonderful overview. The best guide (by far) I've yet read about these folders! Approved (will approve after requested changes are resolved)

CemTunaboylu committed:
minor fixes

CemTunaboylu committed:
rebase me please

CemTunaboylu committed:
I lost count

ozgunozerk reviewed (approved):
Gay

CompositeMan commented:
> Gay

My PR is contaminated.

CemTunaboylu committed:
snap space information deleted

ozgunozerk commented:
Here is the source of `gay`, let this be a reference point to my other approve messages in the future:
![image](https://user-images.githubusercontent.com/32795992/174666749-ab951589-5a00-4f04-95bf-ec8bcf5fc7f8.png)
